keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1537 @@
+another Pod and/or used by multiple different Pods at the same time (if
+supported by the plugin).
+
+### API
+
+ResourceClaim, ResourceClass and ResourcePlugin are new built-in types
+in a new `cdi.k8s.io/v1alpha1` API group. This was chosen instead of
+using CRDs because core Kubernetes components must interact with them
+and installation of CRDs as part of cluster creation is an unsolved
+problem.
+
+The PodSpec gets extended.
+
+```
+type ResourceClass struct {
+    // Resource plugins have a unique name in reverse domain order (acme.example.com).
+    PluginName string
+    // Parameters holds arbitrary values that will be available to the plugin
+    // when allocating a resource that uses this class. The plugin will
+    // be able to distinguish between parameters stored here and
+    // and those stored in ResourceClaimSpec. These parameters
+    // here can only be set by cluster administrators.
+    Parameters runtime.RawExtension
+}
+
+type ResourcePlugin struct {
+    // The name of the object is the unique plugin name.
+    ObjectMeta
+
+    // Features contains a list of features supported by the plugin.
+    // New features may be added over time and must be ignored
+    // by code that does not know about them.
+    Features []ResourcePluginFeature
+}
+
+type ResourcePluginFeature struct {
+    // Name is one of the pre-defined names for a feature.
+    Name ResourcePluginFeatureName
+    // Parameters might provide additional information about how
+    // the plugin supports the feature. Boolean features have
+    // no parameters, merely listing them indicates support.
+    Parameters runtime.RawExtension
+}
+
+type ResourceClaim struct {
+    // The plugin must set a finalizer here before it attempts to
+    // allocate the resource. It removes the finalizer again when
+    // a) the allocation attempt has definitely failed or b) when
+    // the allocated resource was freed. This ensures that
+    // resources are not leaked.
+    ObjectMeta
+    // Spec describes the desired attributes of a resource that then
+    // needs to be allocated. It can only be set once when creating
+    // the ResourceClaim.
+    Spec ResourceClaimSpec
+    // Status describes whether the resource is available and with
+    // which attributes.
+    Status ResourceClaimStatus
+}
+
+type ResourceClaimSpec struct {
+    // ResourceClassName references the plugin and additional
+    // parameters via the name of a ResourceClass that was
+    // created as part of the plugin deployment.
+    //
+    // The apiserver does not check that the referenced class
+    // exists, but a plugin-specific admission webhook
+    // may require that and is allowed to reject claims where
+    // the class is missing.
+    ResourceClassName string
+
+    // Parameters holds arbitrary values that will be available to the plugin
+    // when allocating a resource for the claim.
+    Parameters runtime.RawExtension
```

Comment:
Could you clarify the usgae of these different Parameters?

Reply:
See above. The "Cluster configuration" and "Partial GPU allocation" user story also provide an example.